app/routes.py

```diff
@@ -1,97 +1,232 @@
+                            for org in organizations]
+    form.organization_id.choices = organization_choices
+    
+    if request.is_json:
+        org = db.add_harvest_source(request.json)
+        if org:
+            return jsonify({"message": f"Added new harvest source with ID: {org.id}"})
+        else:
+            return jsonify({"error": "Failed to add harvest source."}), 400
+    else:
+        if form.validate_on_submit():
+            new_source = {
+                "name": form.name.data,
+                "notification_emails": form.emails.data.replace('\r\n', ', '),
+                "frequency": form.frequency.data,
+                "user_requested_frequency": form.frequency.data,
+                "url": form.url.data,
+                "schema_type": form.schema_type.data,
+                "source_type": form.source_type.data,
+                "organization_id": form.organization_id.data
+            }
+            source=db.add_harvest_source(new_source)
+            if source:
+                return f"Added new source with ID: {source.id}"
+            else:
+                return "Failed to add harvest source."
+    return render_template("source_form.html", form=form, choices=organization_choices)
+
+# test interface, will remove later
+@mod.route("/get_harvest_source", methods=["GET"]) 
 def get_all_harvest_sources():
-    result = db.get_all_harvest_sources()
+    source = db.get_all_harvest_sources()
+    org = db.get_all_organizations()
+    return render_template("harvest_source.html", sources=source, organizations=org)
+
+@mod.route("/harvest_source/", methods=["GET"])    
+@mod.route("/harvest_source/<source_id>", methods=["GET"])
+def get_harvest_source(source_id=None):
+    if source_id:
+        source = db.get_harvest_source(source_id)
+        return jsonify(source) if source  else ("Not Found", 404)
+
+    organization_id = request.args.get("organization_id")
+    if organization_id:
+        source = db.get_harvest_source_by_org(organization_id)
+        if not source:
+            return "No harvest sources found for this organization", 404
+    else:
+        source = db.get_all_harvest_sources()
+    return jsonify(source)
+        
+@mod.route("/harvest_source/<source_id>", methods=["PUT"])
+def update_harvest_source(source_id):
+    result = db.update_harvest_source(source_id, request.json)
     return result
 
-@mod.route('/harvest_jobs', methods=['GET'])
-def get_all_harvest_jobs():
-    result = db.get_all_harvest_jobs()
+@mod.route("/harvest_source/<source_id>", methods=["DELETE"])
+def delete_harvest_source(source_id):
+    result = db.delete_harvest_source(source_id)
     return result
 
-@mod.route('/harvest_errors_by_job/<job_id>', methods=['GET'])
-def get_all_harvest_errors_by_job(job_id):
-    try:
-        result = db.get_all_harvest_errors_by_job(job_id)
-        return result
-    except Exception:
-        return " provide job_id"
+@mod.route("/harvest_job/add", methods=["POST"])
```

Comment:
should this be a public route if we're only going to let Flask app add jobs?

Reply:
right, we may make it no-public, and also may protect this with auth.